guava/src/com/google/common/escape/UnicodeEscaper.java

```diff
@@ -242,20 +261,23 @@ protected final String escapeSlow(String s, int index) {
    * @return the Unicode code point for the given index or the negated value of the trailing high
    *     surrogate character at the end of the sequence
    */
-  protected static int codePointAt(CharSequence seq, int index, int end) {
+  @SuppressWarnings("upperbound:argument.type.incompatible")/* highest possible `end` value is `seq.length`,
+  `indexInternal` can't be >= seq.length in while loop.*/
+  protected static int codePointAt(CharSequence seq, @IndexFor("#1") int index, @IndexOrHigh("#1") int end) {
     checkNotNull(seq);
-    if (index < end) {
-      char c1 = seq.charAt(index++);
+    @IndexOrHigh("seq") int indexInternal = index;
+    if (indexInternal < end) {
+      char c1 = seq.charAt(indexInternal++);
       if (c1 < Character.MIN_HIGH_SURROGATE || c1 > Character.MAX_LOW_SURROGATE) {
         // Fast path (first test is probably all we need to do)
         return c1;
       } else if (c1 <= Character.MAX_HIGH_SURROGATE) {
         // If the high surrogate was the last character, return its inverse
-        if (index == end) {
+        if (indexInternal == end) {
           return -c1;
         }
         // Otherwise look for the low surrogate following it
-        char c2 = seq.charAt(index);
+        char c2 = seq.charAt(indexInternal);//(1)
         if (Character.isLowSurrogate(c2)) {
           return Character.toCodePoint(c1, c2);
         }
```

Comment:
It still looks to me as if both `IllegalArgumentException`s below still refer to `index` and not to `indexInternal`